# LittleBrownBats_WNSSurvivorsVsNonsurvivors
Code and input files for Auteri and Knowles 2020, Decimated little brown bats show potential for adaptive change, Scientific Reports (10): 3023.
https://www.nature.com/articles/s41598-020-59797-4

Scripts and other files provided in this repository are intended for understanding and adapting the analyses performed in the publication Auteri, G. and L.L. Knowles, 2019. Please cite this paper when using altered versions of scripts for other projects, with the excpetion of the Filter_SNPs-Loci_Stacks21.R script which requires no citation. The dataset consists of genomic data (Restriction site Associated DNA Sequences; RAD-Seq) for little brown bats (Myotis lucifugus) which are survivors versus non-survivors of the bat disease white-nose syndrome. The objective is to find loci for which the allelic variants are significantly partitioned between disease survivors versus non-survivors (a.k.a. outlier loci), as such loci may be indicative of genes which can confer disease susceptibility versus tolerance/resistance.

###################################################################################
Preliminary data processing / analyses

The Filter_SNPs-Loci_Stacks21.R script takes as input a snps.vcf file (unzip the snps.zip file in this repository; this file is output from the 'populations' step of STACKS) and generates a whitelist of acceptable SNPs and loci (to be used as input for an additional run of 'populations'). Its purpose is to 1) remove loci from consideration that have a suspiciously high per-locus theta value (which may not be artifacts of alignment error versus truely a single locus) and 2) remove variable sites from consideration based on their position in the read (latter sites are particularly prone to sequening errors. The whitelist output by this script is intented to be used as input for an additional run of the 'populations' step in STACKS, as a "loci whitelist." Additional filtering steps can and should be used prior to and after this script.

The Estimate_relatedness.R script uses the R package "related" and the "relatedInput.txt" file in order to estimate relatedness between all pairs of individuals in the dataset. It includes subsections for filtering by missing data and by minor allele frequency, as preliminary analyses suggested the test is sensitive to these factors. These sections can be removed if the user has already performed such filtering steps prior to generating the input txt file. The script then performs simulations to compare various metrics of relatedness for the input dataset, then computes the chosen relatedness metric (in this case Ritland relatedness score) for the empirical data, however caution should be used when evaluting indexes based on simulated data, as they can be misleading. See the "related" tutorial for a more in-depth discussion of these steps.

The PCAScript.R performs a PCA of genomic data, for which the PCA is generated based on survivors. Non-survivors are then plotted onto this PCA--a way of generating a PCA for when the variation in one group of interest is a subset of the variation in another group (e.g. with a founder group versus a source population). The script uses the populations_33Inds_NoHead.stru file as input (a subset of the data, with individuals with large amounts of missing data removed). Loci which are not variable in both groups are then removed (a PCA cannot be calculated if there are some sites with no variation), and filtering is done to remove loci with low minor allele frequency (PCAs can be more sensitive to this compared to some other analyses). The PCAs are then generated and plotted (with different colors indicating survivorship status, and symbols provided for different sampling locations within the study area). 

##########################################################################
Identifying outlier loci

The ProcessFST_STACKS.R script takes the output file from the 'populations' step of STACKS and evaluates outliers based on number of standard deviations from the mean. A plot of FST versus alignment position is output (color coded by reference genome scaffold). It also outputs the SelectedLoci.csv file, which is used as input for the downstream Find_outliers_union.R script.

The Bootstrap_Fst.R script takes as input the populations.fst_mortality-survivor.tsv file (as output from the program STACKS) and assesses 95% confidence intervals on FST values by performing 1,000 bootstraps across loci (using the R package Diversity). The computational requirements exceed that of a personal computer, but for testing putposes one can reduce the number of bootstraps or amount of input data. The script also identifies outliers (defined as loci for which the 95% confidence interval does not fall below five standard deviations from the mean) and writes them to a separate file, in addition to writing the full results to other files. The "outliers" output file is used as input for Find_outliers_union.R.

The Find_Outflank_outliers.R script uses the R package OutFLANK and the OutflankInput.geno, popnamesOutflank.txt, and lociNames.csv input files to identify outlier loci. See the OutFLANK package for more details. The script writes a file (SelectedLoci_OutFlank.csv) which is used as input for Find_outliers_union.R, and also generates several plots.

The Find_outliers_union.R script is meant to take the outputs of three outlier detection methods (the Bootstrap_Fst.R, Find_Outflank_outliers.R, and ProcessFST_STACKS.R scripts) and find the loci for which all three methods agree is a significant outlier. A locus must be identified by all three methods to be included in the SelectedLoci_PreIDs.csv file, whereas any loci identified by just one of the methods is included in the LociOfInterest_PreIDs.csv output.

The ID_Snps_to_Reference.R script takes the SelectedLoci_PreIDs.csv file and looks up information on that genomic site's position using the focal species' reference genome (ref_Myoluc2.0_top_level_reduced.txt). Note that this latter file has been processed somewhat from the file available on the genome (ftp://ftp.ncbi.nih.gov/genomes/Myotis_lucifugus/Gnomon/) so that it can be read into R without issues. The script will loop through each SNP of interest in the SelectedLoci_PreIDs.csv file. If the SNP falls within an annotated region, that information will be saved, but if not, the script will find the nearest annotated region to either side SNP of interest, and store information on those areas. All results are written in the SelectedLociRefIDs_ThreeUnion.csv file.

